### PR TITLE
Fix container restart backoff

### LINKED ISSUE #99697: startup probe restarts without back-off (and looses events)
#### What happened:

- Pods were restarted twice in a row between back-offs:
  - restart
  - restart
  - back-off
  - restart
  - restart
  - back-off
  - restart
  - ...
- Expected events (e.g. `Unhealthy` and `Killing`) did not show the expected count.
  Kubelet log does not show dropped events (also cluster had no other workload).

#### What you expected to happen:

- Back-off occurs after every restart:
  - restart
  - back-off
  - restart
  - back-off
  - restart
- Events count are correct, e.g.:
  - `Killing` events matches the pod restart count
  - `Unhealthy` events match pod restart count times `failureThreshold`

#### How to reproduce it (as minimally and precisely as possible):

`kubectl create ns test`

Apply this:
```yaml
apiVersion: v1
kind: Pod
metadata:
  labels:
    test: liveness
  name: liveness-exec
  namespace: test
spec:
  containers:
  - name: c1
    image: joelnb/dumb-init-alpine # to get proper signal handling
    args:
    - /bin/sh
    - -c
    - sleep 1800
    startupProbe:
      exec:
        command:
        - "false"
      initialDelaySeconds: 20
      periodSeconds: 10
      failureThreshold: 5
```

#### Anything else we need to know?:

1. Events seem to be lost in more cases (like failing readiness probe)
2. Output of describe on GKE:
    ```
    Name:         liveness-exec
    Namespace:    test
    Priority:     0
    Node:         gke-cki-1211-default-pool-7c5dda4f-l962/10.156.0.2
    Start Time:   Wed, 03 Mar 2021 09:21:29 +0100
    Labels:       test=liveness
    Annotations:  <none>
    Status:       Running
    IP:           10.124.0.4
    IPs:
      IP:  10.124.0.4
    Containers:
      c1:
        Container ID:  containerd://eea1660dcec1b233058b268e595b5f73b154a8497a9e7eae83003dffe37e8b25
        Image:         joelnb/dumb-init-alpine
        Image ID:      docker.io/joelnb/dumb-init-alpine@sha256:de41ab5ae9e00d1f3947833a57bd65219af5d35ab8caa4267005b8eeee164f87
        Port:          <none>
        Host Port:     <none>
        Args:
          /bin/sh
          -c
          sleep 1800
        State:          Running
          Started:      Wed, 03 Mar 2021 10:06:21 +0100
        Last State:     Terminated
          Reason:       Error
          Exit Code:    143
          Started:      Wed, 03 Mar 2021 10:00:10 +0100
          Finished:     Wed, 03 Mar 2021 10:01:16 +0100
        Ready:          False
        Restart Count:  14
        Startup:        exec [false] delay=20s timeout=1s period=10s #success=1 #failure=5
        Environment:    <none>
        Mounts:
          /var/run/secrets/kubernetes.io/serviceaccount from default-token-fnkmq (ro)
    Conditions:
      Type              Status
      Initialized       True 
      Ready             False 
      ContainersReady   False 
      PodScheduled      True 
    Volumes:
      default-token-fnkmq:
        Type:        Secret (a volume populated by a Secret)
        SecretName:  default-token-fnkmq
        Optional:    false
    QoS Class:       BestEffort
    Node-Selectors:  <none>
    Tolerations:     node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                     node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
    Events:
      Type     Reason     Age                  From               Message
      ----     ------     ----                 ----               -------
      Normal   Scheduled  46m                  default-scheduler  Successfully assigned test/liveness-exec to gke-cki-1211-default-pool-7c5dda4f-l962
      Normal   Pulled     46m                  kubelet            Successfully pulled image "joelnb/dumb-init-alpine" in 3.14353209s
      Normal   Pulled     43m                  kubelet            Successfully pulled image "joelnb/dumb-init-alpine" in 1.188681438s
      Normal   Pulling    42m (x3 over 46m)    kubelet            Pulling image "joelnb/dumb-init-alpine"
      Normal   Created    42m (x3 over 46m)    kubelet            Created container c1
      Normal   Started    42m (x3 over 46m)    kubelet            Started container c1
      Normal   Pulled     42m                  kubelet            Successfully pulled image "joelnb/dumb-init-alpine" in 1.166587867s
      Normal   Killing    16m (x11 over 43m)   kubelet            Container c1 failed startup probe, will be restarted
      Warning  BackOff    6m4s (x91 over 36m)  kubelet            Back-off restarting failed container
      Warning  Unhealthy  59s (x71 over 45m)   kubelet            Startup probe failed:
    ```
   14 restarts in 46 minutes (see `Scheduled` event) are implausible: The back-off alone should add to about 45 minutes (0s+10s+20s+...+300s+300s+300s) and the pod runs at least 20s (`initialDelaySeconds`) + 50s (`failureThreshold` * `periodSeconds`) per restart. So 14 restarts are only plausible after 1 hour after `Scheduled` event.
    You can also watch this live with `watch kubectl describe pod -n test liveness-exec`.

#### Environment:
- Kubernetes version (use `kubectl version`):
  reproducible on GKE (1.20)
   ```
  Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.4", GitCommit:"e87da0bd6e03ec3fea7933c4b5263d151aafd07c", GitTreeState:"clean", BuildDate:"2021-02-20T02:22:41Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}
  Server Version: version.Info{Major:"1", Minor:"20+", GitVersion:"v1.20.2-gke.2500", GitCommit:"687560e4143e1980f78cd8b8e10e65488203c611", GitTreeState:"clean", BuildDate:"2021-02-17T09:14:18Z", GoVersion:"go1.15.5b5", Compiler:"gc", Platform:"linux/amd64"}
  ```
  also on Gardener (1.20)
  ```  
  Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.4", GitCommit:"e87da0bd6e03ec3fea7933c4b5263d151aafd07c", GitTreeState:"clean", BuildDate:"2021-02-20T02:22:41Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}
  Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.2", GitCommit:"faecb196815e248d3ecfb03c680a4507229c2a56", GitTreeState:"clean", BuildDate:"2021-01-13T13:20:00Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
  ```
  and on Gardener (1.18):
  ```
  Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.4", GitCommit:"e87da0bd6e03ec3fea7933c4b5263d151aafd07c", GitTreeState:"clean", BuildDate:"2021-02-20T02:22:41Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}
  Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.14", GitCommit:"89182bdd065fbcaffefec691908a739d161efc03", GitTreeState:"clean", BuildDate:"2020-12-18T12:02:35Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
  ```

### PR DESCRIPTION
**Pull request by wzshiming**
#### What type of PR is this?
/kind bug
/sig node

#### Which issue(s) this PR fixes:

Fixes #99697

#### Special notes for your reviewer:

<details>
<summary>Pod yaml</summary>

``` yaml
apiVersion: v1
kind: Pod
metadata:
  labels:
    test: liveness
  name: liveness-exec
  namespace: default
spec:
  containers:
    - name: c1
      image: alpine
      imagePullPolicy: IfNotPresent
      args:
        - /bin/sh
        - -c
        - sleep 1800
      startupProbe:
        exec:
          command:
            - "false"
        initialDelaySeconds: 1
        periodSeconds: 1
        failureThreshold: 1
```

</details>

<details>
<summary>Before</summary>

``` 
NAME            READY   STATUS    RESTARTS   AGE
liveness-exec   0/1     Running            0          12s
liveness-exec   0/1     Running            1          34s
liveness-exec   0/1     Running            2          66s
liveness-exec   0/1     Running            3          99s
liveness-exec   0/1     Running            4          2m12s
liveness-exec   0/1     CrashLoopBackOff   4          2m43s
liveness-exec   0/1     Running            5          3m28s
liveness-exec   0/1     CrashLoopBackOff   5          3m59s
liveness-exec   0/1     Running            6          5m32s
liveness-exec   0/1     CrashLoopBackOff   6          6m4s
liveness-exec   0/1     Running            7          8m48s
liveness-exec   0/1     CrashLoopBackOff   7          9m20s
liveness-exec   0/1     Running            8          14m
liveness-exec   0/1     Running            9          15m
liveness-exec   0/1     CrashLoopBackOff   9          15m
liveness-exec   0/1     Running            10         20m
liveness-exec   0/1     Running            11         21m
liveness-exec   0/1     CrashLoopBackOff   11         21m
liveness-exec   0/1     Running            12         26m
liveness-exec   0/1     Running            13         27m
liveness-exec   0/1     CrashLoopBackOff   13         27m
liveness-exec   0/1     Running            14         32m
liveness-exec   0/1     Running            15         33m
liveness-exec   0/1     CrashLoopBackOff   15         33m
liveness-exec   0/1     Running            16         39m
liveness-exec   0/1     Running            17         39m
liveness-exec   0/1     CrashLoopBackOff   17         40m
liveness-exec   0/1     Running            18         45m
liveness-exec   0/1     Running            19         45m
liveness-exec   0/1     CrashLoopBackOff   19         46m
```
</details>

<details>
<summary>After</summary>

``` 
NAME            READY   STATUS    RESTARTS   AGE
liveness-exec   0/1     Running            1          33s
liveness-exec   0/1     Running            2          66s
liveness-exec   0/1     Running            3          98s
liveness-exec   0/1     Running            4          2m11s
liveness-exec   0/1     CrashLoopBackOff   4          2m42s
liveness-exec   0/1     Running            5          2m56s
liveness-exec   0/1     CrashLoopBackOff   5          3m27s
liveness-exec   0/1     Running            6          4m23s
liveness-exec   0/1     CrashLoopBackOff   6          4m54s
liveness-exec   0/1     Running            7          7m8s
liveness-exec   0/1     CrashLoopBackOff   7          7m39s
liveness-exec   0/1     Running            8          12m
liveness-exec   0/1     CrashLoopBackOff   8          12m
liveness-exec   0/1     Running            9          17m
liveness-exec   0/1     CrashLoopBackOff   9          17m
liveness-exec   0/1     Running            10         22m
liveness-exec   0/1     CrashLoopBackOff   10         23m
liveness-exec   0/1     Running            11         27m
liveness-exec   0/1     CrashLoopBackOff   11         28m
```

</details>

#### Does this PR introduce a user-facing change?

```release-note
Fix container restart backoff
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

**Comment by matthyx** [comment]
@wzshiming thanks for the PR... could you please elaborate with `time.Now()` is more suitable than `cStatus.FinishedAt` in this case? What's the rationale behind this choice?

**Comment by wzshiming** [comment]
> @wzshiming thanks for the PR... could you please elaborate why `time.Now()` is more suitable than `cStatus.FinishedAt` in this case? What's the rationale behind this choice?

After the actual test, there was a problem when restarting 3 times, and it could not be restarted anymore. Checking the date found that the test always output `"Back-off restarting failed container" err="back-off 40s restarting failed container=c1 pod=liveness-exec_default(beb742a7-ca09-4744-8bd5-11591edfcf8d)"`

I think it stays at the 40s stage and cannot enter the next stage because cStatus.FinishedAt is fixed

**Comment by matthyx** [comment]
> I think it stays at the 40s stage and cannot enter the next stage because cStatus.FinishedAt is fixed

So maybe the bug is that `cStatus.FinishedAt` isn't reset?

**Comment by wzshiming** [comment]
1. I added test to `IsInBackOffSinceUpdate` and `IsInBackOffSince`, and found that the behavior of `IsInBackOffSince` is a bit strange. And `IsInBackOffSinceUpdate` meets the expected backoff, I first changed `IsInBackOffSince` to `IsInBackOffSinceUpdate`  https://github.com/kubernetes/kubernetes/commit/d20b60018d1cb76ada371ea47611bd720074b4a4#diff-ab6201fe6ab966b661c5a6ea1afb7c27ba57ae1a3b6c558d9d820813c91e2c64R269-R339
2. Then test https://github.com/kubernetes/kubernetes/pull/103257#issuecomment-869631987
3. I checked the ones that currently use `IsInBackOffSinceUpdate` and they all use `time.Now()`. I changed it to `time.Now()`.
4. After testing, the backoff behavior was as expected.

**Comment by matthyx** [comment]
> So maybe the bug is that `cStatus.FinishedAt` isn't reset?

Sorry @wzshiming I still think this is the root of the issue... so not using it actually helps here but doesn't solve the underlying issue IMHO

**Comment by wzshiming** [comment]
Maybe not, If the time passed to `IsInBackOffSinceUpdate` does not change, and `Next` is not called, then the result returned by `IsInBackOffSinceUpdate` will not change. `FinishedAt` of the container remains unchanged until the next container. so it has been in the same time window.



Below is I added debug log and watch pod event. `FinishedAt` has been reset.

``` log
liveness-exec   0/1     Running             0          2s
liveness-exec   0/1     Running             1          35s
liveness-exec   0/1     Running             2          67s
liveness-exec   0/1     CrashLoopBackOff    2          98s
liveness-exec   0/1     Running             3          99s
liveness-exec   0/1     CrashLoopBackOff    3          2m11s
liveness-exec   0/1     Running             4          2m12s
liveness-exec   0/1     CrashLoopBackOff    4          2m43s
```

``` log
Jun 28 22:24:32 zsm-test-1 kubelet[69061]: I0628 22:24:32.508057   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:24:00.207572243 +0800 CST" RestartCount=0
Jun 28 22:25:04 zsm-test-1 kubelet[69061]: I0628 22:25:04.829863   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:24:32.499582066 +0800 CST" RestartCount=1
Jun 28 22:25:04 zsm-test-1 kubelet[69061]: I0628 22:25:04.830025   69061 kuberuntime_manager.go:933] "=============Back-off restarting failed container" err="back-off 10s restarting failed container=c1 pod=liveness-exec_default(f99e437e-b4cf-4d29-afe8-c87276951fdd)"
Jun 28 22:25:04 zsm-test-1 kubelet[69061]: I0628 22:25:04.951014   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:25:04.820234677 +0800 CST" RestartCount=2
Jun 28 22:25:37 zsm-test-1 kubelet[69061]: I0628 22:25:37.076941   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:25:04.820234677 +0800 CST" RestartCount=2
Jun 28 22:25:37 zsm-test-1 kubelet[69061]: I0628 22:25:37.077139   69061 kuberuntime_manager.go:933] "=============Back-off restarting failed container" err="back-off 20s restarting failed container=c1 pod=liveness-exec_default(f99e437e-b4cf-4d29-afe8-c87276951fdd)"
Jun 28 22:25:37 zsm-test-1 kubelet[69061]: I0628 22:25:37.186924   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:25:37.070102815 +0800 CST" RestartCount=3
Jun 28 22:26:09 zsm-test-1 kubelet[69061]: I0628 22:26:09.295399   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:25:37.070102815 +0800 CST" RestartCount=3
Jun 28 22:26:09 zsm-test-1 kubelet[69061]: I0628 22:26:09.295589   69061 kuberuntime_manager.go:933] "=============Back-off restarting failed container" err="back-off 40s restarting failed container=c1 pod=liveness-exec_default(f99e437e-b4cf-4d29-afe8-c87276951fdd)"
Jun 28 22:26:09 zsm-test-1 kubelet[69061]: I0628 22:26:09.432916   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:26:09.290040732 +0800 CST" RestartCount=4
Jun 28 22:26:09 zsm-test-1 kubelet[69061]: I0628 22:26:09.433131   69061 kuberuntime_manager.go:933] "=============Back-off restarting failed container" err="back-off 40s restarting failed container=c1 pod=liveness-exec_default(f99e437e-b4cf-4d29-afe8-c87276951fdd)"
Jun 28 22:26:10 zsm-test-1 kubelet[69061]: I0628 22:26:10.457750   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:26:09.290040732 +0800 CST" RestartCount=4
Jun 28 22:26:10 zsm-test-1 kubelet[69061]: I0628 22:26:10.457939   69061 kuberuntime_manager.go:933] "=============Back-off restarting failed container" err="back-off 40s restarting failed container=c1 pod=liveness-exec_default(f99e437e-b4cf-4d29-afe8-c87276951fdd)"
Jun 28 22:26:23 zsm-test-1 kubelet[69061]: I0628 22:26:23.946299   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:26:09.290040732 +0800 CST" RestartCount=4
Jun 28 22:26:23 zsm-test-1 kubelet[69061]: I0628 22:26:23.946562   69061 kuberuntime_manager.go:933] "=============Back-off restarting failed container" err="back-off 40s restarting failed container=c1 pod=liveness-exec_default(f99e437e-b4cf-4d29-afe8-c87276951fdd)"
Jun 28 22:26:34 zsm-test-1 kubelet[69061]: I0628 22:26:34.947107   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:26:09.290040732 +0800 CST" RestartCount=4
Jun 28 22:26:34 zsm-test-1 kubelet[69061]: I0628 22:26:34.947650   69061 kuberuntime_manager.go:933] "=============Back-off restarting failed container" err="back-off 40s restarting failed container=c1 pod=liveness-exec_default(f99e437e-b4cf-4d29-afe8-c87276951fdd)"
Jun 28 22:26:48 zsm-test-1 kubelet[69061]: I0628 22:26:48.946538   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:26:09.290040732 +0800 CST" RestartCount=4
Jun 28 22:26:48 zsm-test-1 kubelet[69061]: I0628 22:26:48.947070   69061 kuberuntime_manager.go:933] "=============Back-off restarting failed container" err="back-off 40s restarting failed container=c1 pod=liveness-exec_default(f99e437e-b4cf-4d29-afe8-c87276951fdd)"
Jun 28 22:27:01 zsm-test-1 kubelet[69061]: I0628 22:27:01.945863   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:26:09.290040732 +0800 CST" RestartCount=4
Jun 28 22:27:01 zsm-test-1 kubelet[69061]: I0628 22:27:01.946132   69061 kuberuntime_manager.go:933] "=============Back-off restarting failed container" err="back-off 40s restarting failed container=c1 pod=liveness-exec_default(f99e437e-b4cf-4d29-afe8-c87276951fdd)"
Jun 28 22:27:12 zsm-test-1 kubelet[69061]: I0628 22:27:12.946387   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:26:09.290040732 +0800 CST" RestartCount=4
Jun 28 22:27:12 zsm-test-1 kubelet[69061]: I0628 22:27:12.946652   69061 kuberuntime_manager.go:933] "=============Back-off restarting failed container" err="back-off 40s restarting failed container=c1 pod=liveness-exec_default(f99e437e-b4cf-4d29-afe8-c87276951fdd)"
Jun 28 22:27:24 zsm-test-1 kubelet[69061]: I0628 22:27:24.946833   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:26:09.290040732 +0800 CST" RestartCount=4
Jun 28 22:27:24 zsm-test-1 kubelet[69061]: I0628 22:27:24.947091   69061 kuberuntime_manager.go:933] "=============Back-off restarting failed container" err="back-off 40s restarting failed container=c1 pod=liveness-exec_default(f99e437e-b4cf-4d29-afe8-c87276951fdd)"
Jun 28 22:27:35 zsm-test-1 kubelet[69061]: I0628 22:27:35.945933   69061 kuberuntime_manager.go:925] "======DEBUG" FinishedAt="2021-06-28 22:26:09.290040732 +0800 CST" RestartCount=4
Jun 28 22:27:35 zsm-test-1 kubelet[69061]: I0628 22:27:35.946231   69061 kuberuntime_manager.go:933] "=============Back-off restarting failed container" err="back-off 40s restarting failed container=c1 pod=liveness-exec_default(f99e437e-b4cf-4d29-afe8-c87276951fdd)"
```

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go
Comment (matthyx) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go:937]: for consistency, since `backoff.Next()` also checks expiration you should use `now` I guess

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go
Comment (matthyx) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go:926]: ok, changing to `IsInBackOffSinceUpdate` makes it working

**Comment by matthyx** [comment]
> Maybe not, If the time passed to `IsInBackOffSinceUpdate` does not change, and `Next` is not called, then the result returned by `IsInBackOffSinceUpdate` will not change. `FinishedAt` of the container remains unchanged until the next container. so it has been in the same time window.

Yes, got it now... sorry for the delay in understanding. I think indeed it's right to check with now, pls check my inline comments.

**Comment by wzshiming** [comment]
@matthyx  Thanks, updated.

**Comment by matthyx** [comment]
I think it's correct now... let's see what the CI says...
/lgtm

**Comment by liggitt** [comment]
Is the issue specific to failing startup probes? Are existing crash-loop scenarios where the container actually finishes and crashes working properly? If so, can we scope the fix more narrowly so that it doesn't change behavior for existing scenarios that are working?

**Comment by wzshiming** [comment]
> Is the issue specific to failing startup probes?

All kinds of fail have this issue.

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go
Comment (liggitt) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go:925]: this needs a test demonstrating the issue this is fixing

**Comment by liggitt** [comment]
/hold

explicitly holding since it is not clear this is correct, and there is no test demonstrating the issue and the fix

**Comment by wzshiming** [comment]
@liggitt 
I added the e2e test to cover the case of container restart backoff

``` log
Jul  9 20:39:41.075: INFO: container restart backoff: count:1 interval:2s startdAt:2021-07-09 20:39:30 +0800 CST
Jul  9 20:39:45.101: INFO: container restart backoff: count:2 interval:10s startdAt:2021-07-09 20:39:40 +0800 CST
Jul  9 20:40:17.238: INFO: container restart backoff: count:3 interval:4s startdAt:2021-07-09 20:39:44 +0800 CST
Jul  9 20:41:01.443: INFO: container restart backoff: count:4 interval:32s startdAt:2021-07-09 20:40:16 +0800 CST
Jul  9 20:42:30.791: INFO: container restart backoff: count:5 interval:45s startdAt:2021-07-09 20:41:01 +0800 CST
Jul  9 20:45:21.439: INFO: container restart backoff: count:6 interval:89s startdAt:2021-07-09 20:42:30 +0800 CST
Jul  9 20:50:25.635: INFO: container restart backoff: count:7 interval:171s startdAt:2021-07-09 20:45:21 +0800 CST
Jul  9 20:55:33.953: INFO: container restart backoff: count:8 interval:304s startdAt:2021-07-09 20:50:25 +0800 CST
Jul  9 21:00:30.115: INFO: container restart backoff: count:9 interval:307s startdAt:2021-07-09 20:55:32 +0800 CST
```

**Comment by wzshiming** [comment]
@liggitt Hi. could you have time to take a look

**Comment by matthyx** [comment]
Still LGTM, maybe we can merge this after code thaw @mrunalp ?
/lgtm

**Comment by derekwaynecarr** [comment]
I want to review this more closely.  Is it possible to split this PR into a test case that identifies the problem (and consistently fails) separate from the solution?

File: test/e2e_node/container_restart_backoff_test.go
Comment (derekwaynecarr) [inline test/e2e_node/container_restart_backoff_test.go:30]: a separate pr with just this test and a skip on it is helpful that shows how back-off is consistently (or inconsistently) broken.  we have also made some related refactors in kubelet at end of this release where it would be good to accumulate signal on this test in isolation of a fix multiple times.

File: test/e2e_node/container_restart_backoff_test.go
Comment (liggitt) [inline test/e2e_node/container_restart_backoff_test.go:115]: even a 50% tolerance on a 10 second interval still seems like it could be flaky if the kubelet reconcile loop is involved

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go
Comment (derekwaynecarr) [reply pkg/kubelet/kuberuntime/kuberuntime_manager.go:952]: I do not understand the reasoning for no longer looking at the finished timestamp of the last exited container versus the present timestamp, how did that change fix the core issue?  can you elaborate?

File: staging/src/k8s.io/client-go/util/flowcontrol/backoff_test.go
Comment (derekwaynecarr) [inline staging/src/k8s.io/client-go/util/flowcontrol/backoff_test.go:198]: this test seems distinct from the kuberuntime change, and may be useful providing in its own PR.

i am still confused on the usage of the now timestamp though versus last observed exit time.

File: test/e2e_node/container_restart_backoff_test.go
Comment (derekwaynecarr) [reply test/e2e_node/container_restart_backoff_test.go:115]: agree with @liggitt that this could be flaky when kubelet is running large numbers of concurrent pods.  if i read the original reported issue, it looked like there were periods between restarts with no backoff, is it better to just test that?  from a cluster user perspective, we make now outward commitments on retry and backoff periods generally, its more to preserve the health of the node and node agents than for the end-user benefit.

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go
Comment (ehashman) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go:952]: This doesn't seem right to me. We reset the time every run through this function, so won't that prevent an exponential back off? According to the output in the first comment, the restart interval is linear/fixed now, which I believe is a change in behaviour.

The root issue seems to be that there's some sort of race described in #99697 resulting in two restarts. I'm not sure how this change addresses that.

/hold

**Review by ehashman — CHANGES_REQUESTED** [review]

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go
Comment (matthyx) [reply pkg/kubelet/kuberuntime/kuberuntime_manager.go:952]: If you compare `backOff.IsInBackOffSince(key, ts)`
https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/client-go/util/flowcontrol/backoff.go#L92-L103
with `backOff.IsInBackOffSinceUpdate(key, now)`
https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/client-go/util/flowcontrol/backoff.go#L106-L117

You can see the only difference is on the condition `hasExpired`, so does it really changes the exponential backoff?
This might eventually come from the change that I requested when calling `backOff.Next(key, now)`.

I think we should be asking ourselves why there is a bug on the condition described above... also this func in only called from here which would make it easier to fix without side effects.

Also this calls for more investigation:

> Maybe not, If the time passed to IsInBackOffSinceUpdate does not change, and Next is not called, then the result returned by IsInBackOffSinceUpdate will not change. FinishedAt of the container remains unchanged until the next container. so it has been in the same time window.

Which is what we see happening...

File: staging/src/k8s.io/client-go/util/flowcontrol/backoff_test.go
Comment (wzshiming) [reply staging/src/k8s.io/client-go/util/flowcontrol/backoff_test.go:198]: I sent a new PR https://github.com/kubernetes/kubernetes/pull/104139

I have the same confusion about `IsInBackOffSince`. Fortunately, there is no other place to use it except here. `IsInBackOffSinceUpdate` is used in the same way in other places. After that, we should use a clearer `IsInBackOffSinceUpdate`

File: test/e2e_node/container_restart_backoff_test.go
Comment (wzshiming) [reply test/e2e_node/container_restart_backoff_test.go:30]: OK, I sent a new PR https://github.com/kubernetes/kubernetes/pull/104140

**Comment by wzshiming** [comment]
https://github.com/kubernetes/kubernetes/pull/103257#discussion_r682667876
https://github.com/kubernetes/kubernetes/pull/103257#discussion_r682679708

Split test into #104139 #104140

**Comment by wzshiming** [comment]
@derekwaynecarr 

Hi, could you have time to take a look

**Comment by wzshiming** [comment]
@derekwaynecarr HI, Could you have time to take a look?
This PR does fix the backoff strategy for container restarts. How can I forward it?

**Comment by wzshiming** [comment]
@ehashman Hi

Could you have time to look again? This is indeed a bug in the restart strategy and needs to be fixed.

**Comment by wzshiming** [comment]
friendly ping @derekwaynecarr

**Pull request closed without merging**
